### PR TITLE
[ioctl/newcmd] build account sign command line into new ioctl

### LINKED ISSUE #3001: [ioctl] build `account sign` command line into new ioctl
The new ioctl (./ioctl/newcmd) is a refactored version of the old ioctl(./ioctl/cmd). It includes two major changes:
- The [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) is passed when constructing the Cobra command.
- Unit tests are to be added for each command.

The newcmd code currently is a draft. The code of old ioctl can be used for reference.

### What would you like to be added
add `accountsign.go` and `accountsign_test.go` in `./ioctl/newcmd/account`. 
**Code for reference**:
https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/account/accountsign.go

### PR DESCRIPTION
**Pull request by mmyj**
close #3001 
Add `accountsign.go` and `accountsign_test.go` in `./ioctl/newcmd/account`.

### CODE REVIEW

**Review by Liuhaai — APPROVED** [review]

File: ioctl/newcmd/account/accountsign_test.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/iotexproject/iotex-core/test/mock/mock_ioctlclient"
+)
+
+func TestNewAccountSign(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	client := mock_ioctlclient.NewMockClient(ctrl)
+	client.EXPECT().SelectTranslation(gomock.Any()).Return("mockTranslationString", config.English).AnyTimes()
+
+	// test for not invalid account address
```
Comment (Liuhaai) [inline ioctl/newcmd/account/accountsign_test.go:26]: could you add a valid test?

Comment (mmyj) [reply ioctl/newcmd/account/accountsign_test.go:26]: The test will be blocked to wait for password when invoke `util.ReadSecretFromStdin()`, have any solution?

Comment (mmyj) [reply ioctl/newcmd/account/accountsign_test.go:26]: This is the test code
```
// test for exist account
testWallet := filepath.Join(os.TempDir(), "testNewAccountSign")
defer testutil.CleanupPath(t, testWallet)

originWallet := config.ReadConfig.Wallet
config.ReadConfig.Wallet = testWallet
defer func() {
config.ReadConfig.Wallet = originWallet
}()
ks := keystore.NewKeyStore(config.ReadConfig.Wallet, keystore.StandardScryptN, keystore.StandardScryptP)
require.NotNil(ks)

// create accounts
cmd = NewAccountSign(client)
account, err := ks.NewAccount("test")
require.NoError(err)
addr, err := address.FromBytes(account.Address.Bytes())
require.NoError(err)
require.NoError(cmd.Flag("signer").Value.Set(addr.String()))
result, err := util.ExecuteCmd(cmd)
require.NoError(err)
require.Equal("", result)
```

**Comment by mmyj** [comment]
I had tried this unit test
```
func TestReadSecretFromStdin(t *testing.T) {
	_, err := util.ReadSecretFromStdin()
	require.NoError(t, err)
}
```

and get an error 
```
  accountsign_test.go:34: 
       Error Trace:	accountsign_test.go:34
       Error:      	Received unexpected error:
        	           : operation not supported by device
       Test:       	TestReadSecretFromStdin
```

I'm not found the reason for this error. Do you have any ideas?

File: ioctl/newcmd/account/accountsign_test.go (hunk 1 above)
Comment (Liuhaai) [reply ioctl/newcmd/account/accountsign_test.go:26]: > 

`ReadSecret()` in client can be used for this pr. And use mocked one in the test

Comment (mmyj) [reply ioctl/newcmd/account/accountsign_test.go:26]: Should I refactor the `ioctl/newcmd/account/account.Sign()`, let the new implement read password by `client.ReadSecret()`?

Comment (Liuhaai) [reply ioctl/newcmd/account/accountsign_test.go:26]: sure

Comment (mmyj) [reply ioctl/newcmd/account/accountsign_test.go:26]: Make sense of it.

Comment (mmyj) [reply ioctl/newcmd/account/accountsign_test.go]: Done, PTAL.

**Comment by huof6829** [comment]
> I had tried this unit test
> 
> ```
> func TestReadSecretFromStdin(t *testing.T) {
> 	_, err := util.ReadSecretFromStdin()
> 	require.NoError(t, err)
> }
> ```
> 
> and get an error
> 
> ```
>   accountsign_test.go:34: 
>        Error Trace:	accountsign_test.go:34
>        Error:      	Received unexpected error:
>         	           : operation not supported by device
>        Test:       	TestReadSecretFromStdin
> ```
> 
> I can not find the reason for this error. Do you have any ideas?

this need user input when run.  use  client.ReadSecret.  and test the result by gomock.

**Comment by mmyj** [comment]
> 

Thanks for your suggesting. I had been refactored the `account.Sign()` same as your suggestion, it's invoke `client.ReadSecret` now.

**Comment by huof6829** [comment]
i run go test -v ioctl/newcmd/account/...  and found Fail in TestNewAccountSign.  pls check

--- FAIL: TestNewAccountSign (1.62s)
    accountsign_test.go:52:
        	Error Trace:	accountsign_test.go:52
        	Error:      	Received unexpected error:
        	            	failed to sign message: invalid address #io16j6syjlwr3crsxh879lr9l0c7um64pathvzx6y
        	Test:       	TestNewAccountSign
    --- FAIL: TestNewAccountSign/valid_account (1.62s)
        testing.go:1169: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
FAIL
FAIL	github.com/iotexproject/iotex-core/ioctl/newcmd/account	6.832s
FAIL

**Comment by Liuhaai** [comment]
> i run go test -v ioctl/newcmd/account/... and found Fail in TestNewAccountSign. pls check
> 
> --- FAIL: TestNewAccountSign (1.62s) accountsign_test.go:52: Error Trace: accountsign_test.go:52 Error: Received unexpected error: failed to sign message: invalid address #io16j6syjlwr3crsxh879lr9l0c7um64pathvzx6y Test: TestNewAccountSign --- FAIL: TestNewAccountSign/valid_account (1.62s) testing.go:1169: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test FAIL FAIL github.com/iotexproject/iotex-core/ioctl/newcmd/account 6.832s FAIL

Could you try pull the lastest code

**Comment by mmyj** [comment]
> i run go test -v ioctl/newcmd/account/...  and found Fail in TestNewAccountSign.  pls check
> 
> 
> 
> --- FAIL: TestNewAccountSign (1.62s)
> 
>     accountsign_test.go:52:
> 
>         	Error Trace:	accountsign_test.go:52
> 
>         	Error:      	Received unexpected error:
> 
>         	            	failed to sign message: invalid address #io16j6syjlwr3crsxh879lr9l0c7um64pathvzx6y
> 
>         	Test:       	TestNewAccountSign
> 
>     --- FAIL: TestNewAccountSign/valid_account (1.62s)
> 
>         testing.go:1169: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
> 
> FAIL
> 
> FAIL	github.com/iotexproject/iotex-core/ioctl/newcmd/account	6.832s
> 
> FAIL

I will check it asap.

**Comment by jrynkiew** [comment]
Should messages even be allowed to be signed without a prefix? It's open to exploits if it just signs, right?

**Comment by mmyj** [comment]
> go test -v ioctl/newcmd/account/...

All tests are passed in my local. And git log commit is `97d02599f75f3f0a0aab7c6bdc5b791f36f65c4f`.
```
~/sdk/gopath/src/github.com/mmyj/iotex-core/ioctl/newcmd/account$ go test -v .                   
=== RUN   TestNewAccountCreate
{
  "accounts": [
    {
      "address": "io1zsyelw2x9ct2x3dpn5ppzhm0uc82fmdhv8amnu",
      "privateKey": "27f806a6ad7f3ce4b24f47000d5188e3a61647ac13050b856b1688cf351b2ef3",
      "publicKey": "0411d6cefa6a1f7ddb786314b04a0c9681e628e1124b6acf2c9958e03198819984aea9bbabe38004cd265e48abfbef5a23a94d157501e255ad11c60d40f53ff08e"
    }
  ]
}
--- PASS: TestNewAccountCreate (0.00s)
=== RUN   TestNewAccountDelete
quit
mockTranslationString%!(EXTRA string=io18wgj7u9vj8ngpxhjhwzslg55std5aktwk6jwx6)
--- PASS: TestNewAccountDelete (1.51s)
=== RUN   TestNewAccountList
=== RUN   TestNewAccountList/When_NewAccountList_returns_no_error
io1wnf7qnt97khz3nl7nq4k0qt2dsml0cx3qlmpgk
io1vdf6xqd2j6nlph6klar9zap097kryvluz3q478
=== RUN   TestNewAccountList/When_NewAccountList_returns_error
--- PASS: TestNewAccountList (2.85s)
    --- PASS: TestNewAccountList/When_NewAccountList_returns_no_error (2.85s)
    --- PASS: TestNewAccountList/When_NewAccountList_returns_error (0.00s)
=== RUN   TestNewAccountNonce
io1rc2d2de7rtuucalsqv4d9ng0h297t63w7wvlph:
Nonce: 0, Pending Nonce: 0
io1rc2d2de7rtuucalsqv4d9ng0h297t63w7wvlph:
Nonce: 0, Pending Nonce: 1
io1rc2d2de7rtuucalsqv4d9ng0h297t63w7wvlph:
Nonce: 2, Pending Nonce: 3
--- PASS: TestNewAccountNonce (0.00s)
=== RUN   TestNewAccountSign
=== RUN   TestNewAccountSign/invalid_account
=== RUN   TestNewAccountSign/valid_account
Enter password for #io14gp8yz93x6qcsj4rwffdfcc7h20q5glnkhajle:

0cb5076a0ce528020d00d86c559af30946d0844f03a1aad6fd37f374a8e34e9b52dde39e27f70fb33462f274c0115a04309680600fe8fcc46bd6742499d0805101
--- PASS: TestNewAccountSign (2.07s)
    --- PASS: TestNewAccountSign/invalid_account (0.00s)
    --- PASS: TestNewAccountSign/valid_account (2.07s)
PASS
ok      github.com/iotexproject/iotex-core/ioctl/newcmd/account 6.814s

```

**Comment by mmyj** [comment]
> 
According to the implement of `account.Sign()`, the message will add a prefix before signing like this.
```.
prefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(b))
```
So I don't quite understand abort you say that `Should messages even be allowed to be signed without a prefix?`.
Could you please explain it in detail? Thanks.

**Comment by mmyj** [comment]
> Should messages even be allowed to be signed without a prefix? It's open to exploits if it just signs, right? This is what I'm talking about, maybe it's relevant to consider signing with "IoTeX signed message" or something like that? https://web3js.readthedocs.io/en/1.0/web3-eth-accounts.html#sign

According to the implement of `account.Sign()`, the message will add a prefix before signing like this.
```.
prefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(b))
```
So I don't quite understand abort you say that `Should messages even be allowed to be signed without a prefix?`.
Could you please explain it in detail? Thanks.

**Comment by jrynkiew** [comment]
> According to the implement of `account.Sign()`, the message will add a prefix before signing like this.
> 
> ```
> prefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(b))
> ```
> 
> So I don't quite understand abort you say that `Should messages even be allowed to be signed without a prefix?`. Could you please explain it in detail? Thanks.

Sorry, I didn't read carefully, I thought you wanted to implement a un-prefixed sign method. Don't worry

**Comment by mmyj** [comment]
> > According to the implement of `account.Sign()`, the message will add a prefix before signing like this.
> > ```
> > prefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(b))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So I don't quite understand abort you say that `Should messages even be allowed to be signed without a prefix?`. Could you please explain it in detail? Thanks.
> 
> Sorry, I didn't read carefully, I thought you wanted to implement a un-prefixed sign method. Don't worry

:ok_hand:

**Comment by huof6829** [comment]
i update the latest code, run ok

**Pull request merged**